File: python/fundamentals/VerifyBST.py
import Traversal

class VerifyBST:
    def __init__(self, root):
        self.root = root

    def verifyBST(self):
        """
        Verify if the tree is a valid BST.
        Returns True if valid, False otherwise.
        """
        def isValidBST(node, min_val, max_val):
            if not node:
                return True
            if node.data <= min_val or node.data >= max_val:
                return False
            return (isValidBST(node.left, min_val, node.data) and
                    isValidBST(node.right, node.data, max_val))

        # Verify BST property
        is_valid = isValidBST(self.root, float('-inf'), float('inf'))

        # Print in-order traversal
        traversal = Traversal.Traversal()
        in_order = traversal.inOrder(self.root)
        print("In-order traversal:", " ".join(map(str, in_order)))

        # Confirm in-order is sorted
        is_sorted = all(in_order[i] <= in_order[i + 1] for i in range(len(in_order) - 1)) if in_order else True

        return is_valid and is_sorted

# Testing like Java's main
def main():
    # Test Case 1: Valid BST
    #       10
    #      /  \
    #     5    15
    #    / \     \
    #   3   7    18
    root1 = Traversal.Node(10)
    root1.left = Traversal.Node(5)
    root1.right = Traversal.Node(15)
    root1.left.left = Traversal.Node(3)
    root1.left.right = Traversal.Node(7)
    root1.right.right = Traversal.Node(18)

    # Test Case 2: Invalid BST
    #       10
    #      /  \
    #     5    15
    #    / \    /
    #   3   7  12
    root2 = Traversal.Node(10)
    root2.left = Traversal.Node(5)
    root2.right = Traversal.Node(15)
    root2.left.left = Traversal.Node(3)
    root2.left.right = Traversal.Node(7)
    root2.right.left = Traversal.Node(12)  # Violates BST: 12 < 15 but in left-subtree

    # Run tests
    print("Test Case 1 (Valid BST):")
    bst1 = VerifyBST(root1)
    print(f"Is valid BST? {bst1.verifyBST()}\n")

    print("Test Case 2 (Invalid BST):")
    bst2 = VerifyBST(root2)
    print(f"Is valid BST? {bst2.verifyBST()}")

if __name__ == "__main__":
    main()

File: java/fundamentals/oop/traversal_code/VerifyBST.java
package fundamentals.oop.traversal_code;

public class VerifyBST {
    private Traversal.Node root;

    public VerifyBST(Traversal.Node root) {
        this.root = root;
    }

    public boolean verifyBST() {
        // Print in-order traversal using Traversal class
        System.out.print("In-order traversal: ");
        Traversal traversal = new Traversal();
        traversal.setRoot(root);
        traversal.inOrder();

        // Verify BST property using range check
        return isValidBST(root, null, null);
    }

    private boolean isValidBST(Traversal.Node node, String minVal, String maxVal) {
        if (node == null) {
            return true;
        }
        // Check if node.data is within bounds (minVal < node.data < maxVal)
        if (minVal != null && node.data.compareTo(minVal) <= 0) {
            return false;
        }
        if (maxVal != null && node.data.compareTo(maxVal) >= 0) {
            return false;
        }
        return isValidBST(node.left, minVal, node.data) && isValidBST(node.right, node.data, maxVal);
    }

    public static void main(String[] args) {
        // Test Case 1: Valid BST
        //       "10"
        //      /    \
        //     "5"   "15"
        //    / \      \
        //   "3" "7"  "18"
        Traversal.Node root1 = new Traversal.Node("10");
        root1.left = new Traversal.Node("5");
        root1.right = new Traversal.Node("15");
        root1.left.left = new Traversal.Node("3");
        root1.left.right = new Traversal.Node("7");
        root1.right.right = new Traversal.Node("18");

        // Test Case 2: Invalid BST
        //       "10"
        //      /    \
        //     "5"   "15"
        //    / \     /
        //   "3" "7" "12"
        Traversal.Node root2 = new Traversal.Node("10");
        root2.left = new Traversal.Node("5");
        root2.right = new Traversal.Node("15");
        root2.left.left = new Traversal.Node("3");
        root2.left.right = new Traversal.Node("7");
        root2.right.left = new Traversal.Node("12"); // Violates BST: "12" < "15" but in left subtree

        // Run tests
        System.out.println("Test Case 1 (Valid BST):");
        VerifyBST bst1 = new VerifyBST(root1);
        System.out.println("\nIs valid BST? " + bst1.verifyBST());
        System.out.println();

        System.out.println("Test Case 2 (Invalid BST):");
        VerifyBST bst2 = new VerifyBST(root2);
        System.out.println("\nIs valid BST? " + bst2.verifyBST());
    }
}